File: src/commands/status.ts
import { Command } from 'commander';
import chalk from 'chalk';
import { JsonStorageService } from '../services/storage-service';
import { normalizeCode } from '../utils/code-generator';
import { logger } from '../utils/logger';
import * as path from 'path';

export const statusCommand = new Command('status')
  .description('Check the status of a target or session')
  .argument('<code>', 'Target code or session ID to check')
  .action(async (code: string) => {
    try {
      const dataDir = path.join(process.cwd(), 'data');
      const storage = new JsonStorageService(dataDir);

      // First, try to find as session ID
      let session = await storage.getSession(code);

      if (!session) {
        // If not a session ID, search for target code
        const allSessions = await storage.getAllSessions();
        const normalizedCode = normalizeCode(code).toUpperCase();

        for (const s of allSessions) {
          const target = s.targets.find(
            (t) => normalizeCode(t.code).toUpperCase() === normalizedCode
          );
          if (target) {
            session = s;
            // Show specific target status
            showTargetStatus(target, s);
            return;
          }
        }

        console.log(chalk.red(`\nTarget or session not found: ${code}\n`));
        console.log(chalk.gray('Use "glimpse list" to see available sessions and targets.\n'));
        process.exit(1);
      }

      // Show session status
      showSessionStatus(session);
    } catch (error) {
      logger.error('Failed to check status');
      if (error instanceof Error) {
        logger.error(error.message);
      }
      process.exit(1);
    }
  });

function showTargetStatus(target: any, session: any): void {
  const now = new Date();
  const canReveal = now >= session.revealAt;
  const timeUntilReveal = session.revealAt.getTime() - now.getTime();

  console.log(chalk.bold.cyan(`\nTarget Status: ${target.code}\n`));
  console.log(chalk.white(`Status: ${getStatusBadge(target.revealed, canReveal)}`));
  console.log(chalk.gray(`Session: ${session.name || session.id}`));
  console.log(chalk.gray(`Created: ${session.createdAt.toLocaleString()}`));
  console.log(chalk.gray(`Reveal:  ${session.revealAt.toLocaleString()}`));

  if (target.revealed) {
    console.log(chalk.gray(`Revealed: ${target.revealedAt?.toLocaleString()}`));
  } else if (canReveal) {
    console.log(chalk.green('\nReady to reveal!'));
    console.log(chalk.gray(`Use: glimpse reveal ${target.code}\n`));
  } else {
    const minutes = Math.ceil(timeUntilReveal / (1000 * 60));
    const hours = Math.floor(minutes / 60);
    const remainingMinutes = minutes % 60;

    console.log(chalk.yellow(`\nTime remaining: ${hours}h ${remainingMinutes}m`));
    console.log(chalk.gray(`Use "glimpse reveal ${target.code} --force" to reveal now\n`));
  }
}

function showSessionStatus(session: any): void {
  const now = new Date();
  const canReveal = now >= session.revealAt;
  const timeUntilReveal = session.revealAt.getTime() - now.getTime();
  const revealedCount = session.targets.filter((t: any) => t.revealed).length;
  const totalTargets = session.targets.length;

  console.log(chalk.bold.cyan(`\nSession Status\n`));
  console.log(chalk.white(`ID: ${chalk.bold(session.id)}`));
  if (session.name) {
    console.log(chalk.white(`Name: ${session.name}`));
  }
  console.log(chalk.gray(`Created: ${session.createdAt.toLocaleString()}`));
  console.log(chalk.gray(`Reveal:  ${session.revealAt.toLocaleString()}`));
  console.log(chalk.gray(`Targets: ${revealedCount}/${totalTargets} revealed\n`));

  if (canReveal) {
    console.log(chalk.green('Session is ready to reveal!\n'));
  } else {
    const minutes = Math.ceil(timeUntilReveal / (1000 * 60));
    const hours = Math.floor(minutes / 60);
    const remainingMinutes = minutes % 60;
    console.log(chalk.yellow(`Time remaining: ${hours}h ${remainingMinutes}m\n`));
  }

  // List targets
  console.log(chalk.bold('Targets:'));
  session.targets.forEach((target: any, index: number) => {
    const status = getStatusBadge(target.revealed, canReveal);
    console.log(chalk.gray(`  ${index + 1}. ${chalk.cyan(target.code)} - ${status}`));
  });
  console.log('');
}

function getStatusBadge(revealed: boolean, canReveal: boolean): string {
  if (revealed) {
    return chalk.green('REVEALED');
  }
  if (canReveal) {
    return chalk.yellow('READY');
  }
  return chalk.blue('ACTIVE');
}

File: glimpse/commands/status.py
"""Status command for checking sessions"""

import os
from datetime import datetime

import click

from glimpse.services.storage_service import StorageService


@click.command()
def status():
    """Check status of all sessions"""

    data_dir = os.path.join(os.getcwd(), 'data')
    storage = StorageService(data_dir)

    sessions = storage.get_all_sessions()

    if not sessions:
        click.echo(click.style('\nNo sessions found.', fg='yellow'))
        click.echo(click.style('Create one with: glimpse create\n', fg='bright_black'))
        return

    now = datetime.now()

    # Count totals
    total_sessions = len(sessions)
    total_targets = sum(len(s['targets']) for s in sessions)
    revealed_targets = sum(sum(1 for t in s['targets'] if t['revealed']) for s in sessions)

    def is_ready(session):
        reveal_at = datetime.fromisoformat(session['revealAt'])
        if reveal_at.tzinfo is not None:
            reveal_at = reveal_at.replace(tzinfo=None)
        return now >= reveal_at

    ready_sessions = sum(1 for s in sessions if is_ready(s))

    click.echo()
    click.echo(click.style('-' * 50, fg='cyan'))
    click.echo(click.style('  GLIMPSE RV STATUS', fg='cyan', bold=True))
    click.echo(click.style('-' * 50, fg='cyan'))
    click.echo()
    click.echo(f"  Total Sessions: {click.style(str(total_sessions), fg='green', bold=True)}")
    click.echo(f"  Ready to Reveal: {click.style(str(ready_sessions), fg='yellow', bold=True)}")
    click.echo(f"  Total Targets: {click.style(str(total_targets), fg='cyan', bold=True)}")
    click.echo(f"  Revealed: {click.style(str(revealed_targets), fg='green', bold=True)} / Pending: {click.style(str(total_targets - revealed_targets), fg='yellow', bold=True)}")
    click.echo()

    if ready_sessions > 0:
        click.echo(click.style('  Sessions ready to reveal:', fg='green', bold=True))
        click.echo()
        for session in sessions:
            reveal_at = datetime.fromisoformat(session['revealAt'])
            if reveal_at.tzinfo is not None:
                reveal_at = reveal_at.replace(tzinfo=None)
            if now >= reveal_at:
                name = session.get('name') or session['id']
                unrevealed = [t for t in session['targets'] if not t['revealed']]
                if unrevealed:
                    click.echo(f"    * {click.style(name, fg='cyan')}")
                    for target in unrevealed:
                        click.echo(f"      - {click.style(target['code'], fg='yellow')}")
        click.echo()

    click.echo(click.style('-' * 50, fg='cyan'))
    click.echo()
